File: src/okf_devkit/renderer.py
"""OKF Markdown bundle を閲覧用 HTML に変換する決定的レンダラー。

Markdown を正本のまま保持し、Bundle 対象の全ページを HTML にミラーする。
生成 HTML 内の Bundle 内リンクだけを ``.html`` へ書き換えるため、ブラウザで
``file://`` として開いても HTML 間を移動できる。
"""

from __future__ import annotations

import hashlib
import html
import os
import posixpath
import re
import shutil
import tempfile
import unicodedata
from dataclasses import dataclass, field
from pathlib import Path, PurePosixPath
from typing import Any, Callable
from urllib.parse import urlsplit, urlunsplit

try:
    from markdown_it import MarkdownIt
except ImportError:  # pragma: no cover - cmd_render が利用者向けメッセージに変換する
    MarkdownIt = None  # type: ignore[assignment]


ASSET_DIR = Path(__file__).resolve().parent / "assets"
AUTO_MARKER_RE = re.compile(r"^\s*<!--\s*okf:auto:(?:start|end)\s*-->\s*$", re.MULTILINE)
SAFE_TYPE_RE = re.compile(r"[^a-z0-9]+")
TEMPLATE_TOKEN_RE = re.compile(r"\{\{([A-Z0-9_]+)\}\}")
OWNERSHIP_MARKER = "<!-- okf-devkit-owned:v1 -->"


class RenderError(RuntimeError):
    """HTML 生成を完了できない場合のエラー。"""


@dataclass
class Heading:
    level: int
    title: str
    anchor: str


@dataclass
class Page:
    doc: Any
    source_rel: str
    output_rel: str
    tokens: list[Any] = field(default_factory=list)
    headings: list[Heading] = field(default_factory=list)
    body_html: str = ""
    char_count: int = 0
    outbound: set[str] = field(default_factory=set)


@dataclass
class RenderReport:
    pages: int
    written: int
    removed: int
    warnings: list[str]
    output_root: Path


def _asset_text(name: str) -> str:
    path = ASSET_DIR / name
    if not path.exists():
        raise RenderError(f"HTML アセットが見つかりません: {path}")
    return path.read_text(encoding="utf-8").replace("\r\n", "\n")


def _slug(text: str, used: dict[str, int]) -> str:
    normalized = unicodedata.normalize("NFKC", text).strip().lower()
    normalized = re.sub(r"[`*_~\[\](){}<>]", "", normalized)
    normalized = re.sub(r"[^\w\-\u3040-\u30ff\u3400-\u9fff]+", "-", normalized, flags=re.UNICODE)
    normalized = normalized.strip("-_") or "section"
    count = used.get(normalized, 0) + 1
    used[normalized] = count
    return normalized if count == 1 else f"{normalized}-{count}"


def _iter_tokens(tokens: list[Any]):
    for token in tokens:
        yield token
        children = getattr(token, "children", None)
        if children:
            yield from _iter_tokens(children)


def _type_class(type_name: str) -> str:
    value = SAFE_TYPE_RE.sub("-", type_name.lower()).strip("-")
    return value or "document"


def _trust_level(doc: Any) -> tuple[str, str]:
    verified = doc.get("verified", [])
    if not isinstance(verified, list) or not verified:
        return "unverified", "未検証"
    for entry in verified:
        if isinstance(entry, dict) and str(entry.get("by", "")).startswith("human:"):
            return "human-reviewed", "人間レビュー済み"
    return "machine-confirmed", "機械確認済み"


def _normalize_bundle_target(source_rel: str, href_path: str) -> str | None:
    """href のパスを Bundle ルート相対の POSIX パスへ解決する。"""
    if href_path.startswith("/"):
        candidate = href_path.lstrip("/")
    else:
        candidate = posixpath.join(posixpath.dirname(source_rel), href_path)
    if href_path.endswith("/"):
        candidate = posixpath.join(candidate, "index.md")
    candidate = posixpath.normpath(candidate)
    if candidate == ".." or candidate.startswith("../"):
        return None
    return candidate.lstrip("./")


def rewrite_href(
    source_rel: str,
    output_rel: str,
    href: str,
    page_sources: set[str],
) -> tuple[str, str | None, str | None]:
    """Bundle 内 Markdown リンクを生成 HTML への相対リンクに変換する。"""
    if not href or href.startswith("#") or href.startswith("//"):
        return href, None, None
    parsed = urlsplit(href)
    if parsed.scheme or not parsed.path:
        return href, None, None
    if not (parsed.path.lower().endswith(".md") or parsed.path.endswith("/")):
        return href, None, None

    target = _normalize_bundle_target(source_rel, parsed.path)
    if target is None:
        return href, None, f"{source_rel}: Bundle 外を指すリンクを保持しました: {href}"
    if target not in page_sources:
        return href, None, f"{source_rel}: HTML 化できないリンク先を保持しました: {href}"

    target_html = str(PurePosixPath(target).with_suffix(".html"))
    output_dir = posixpath.dirname(output_rel) or "."
    relative = posixpath.relpath(target_html, output_dir)
    rewritten = urlunsplit(("", "", relative, parsed.query, parsed.fragment))
    return rewritten, target, None


def _markdown() -> Any:
    if MarkdownIt is None:
        raise RenderError(
            "markdown-it-py が必要です。"
            " `pip install okf-devkit` を実行してください。"
        )
    md = MarkdownIt("commonmark", {"html": False, "linkify": False, "typographer": False})
    md.enable("table")
    default_fence = md.renderer.rules.get("fence")

    def render_fence(tokens, idx, options, env):
        token = tokens[idx]
        language = token.info.strip().split(maxsplit=1)[0] if token.info.strip() else ""
        if language.lower() == "mermaid":
            source = html.escape(token.content, quote=False)
            return (
                '<figure class="diagram"><div class="mermaid" data-mermaid-source="true">'
                f"{source}</div><figcaption>Mermaid diagram</figcaption></figure>\n"
            )
        rendered = default_fence(tokens, idx, options, env) if default_fence else ""
        if token.content.count("\n") > 40:
            label = html.escape(language or "code")
            return (
                '<details class="long-code"><summary>'
                f"長いコードを表示（{label}）</summary>{rendered}</details>\n"
            )
        return rendered

    md.renderer.rules["fence"] = render_fence
    md.renderer.rules["table_open"] = lambda *_args: '<div class="table-scroll"><table>\n'
    md.renderer.rules["table_close"] = lambda *_args: "</table></div>\n"
    return md


def _prepare_page(page: Page, md: Any) -> None:
    source = AUTO_MARKER_RE.sub("", page.doc.body)
    page.tokens = md.parse(source)
    used: dict[str, int] = {}
    text_parts: list[str] = []
    for idx, token in enumerate(page.tokens):
        if token.type == "heading_open" and idx + 1 < len(page.tokens):
            inline = page.tokens[idx + 1]
            title = getattr(inline, "content", "").strip()
            level = int(token.tag[1:]) if token.tag.startswith("h") else 2
            anchor = _slug(title, used)
            token.attrSet("id", anchor)
            page.headings.append(Heading(level, title, anchor))
        if token.type in {"inline", "fence", "code_block"}:
            text_parts.append(getattr(token, "content", ""))
    page.char_count = len(re.sub(r"\s+", "", "".join(text_parts)))


def _rewrite_page_links(page: Page, page_sources: set[str], warnings: list[str]) -> None:
    for token in _iter_tokens(page.tokens):
        if token.type != "link_open":
            continue
        href = token.attrGet("href")
        if not href:
            continue
        rewritten, target, warning = rewrite_href(
            page.source_rel, page.output_rel, href, page_sources
        )
        token.attrSet("href", rewritten)
        if target:
            page.outbound.add(target)
        if warning:
            token.attrJoin("class", "is-broken")
            warnings.append(warning)


def _render_toc(headings: list[Heading]) -> str:
    visible = [heading for heading in headings if heading.level >= 2]
    if len(visible) < 3:
        return ""
    items = []
    for heading in visible:
        indent = max(0, heading.level - 2)
        items.append(
            f'<li class="toc-level-{indent}"><a href="#{html.escape(heading.anchor, quote=True)}">'
            f"{html.escape(heading.title)}</a></li>"
        )
    return '<nav class="toc" aria-label="ページ内目次"><h2>目次</h2><ol>' + "".join(items) + "</ol></nav>"


def _relative_html(from_output: str, target_source: str) -> str:
    target = str(PurePosixPath(target_source).with_suffix(".html"))
    return posixpath.relpath(target, posixpath.dirname(from_output) or ".")


def _render_navigation(pages: list[Page], current: Page) -> str:
    groups: dict[str, list[Page]] = {}
    for page in pages:
        parts = PurePosixPath(page.source_rel).parts
        group = parts[0] if len(parts) > 1 else "ルート"
        groups.setdefault(group, []).append(page)
    order = ["ルート", "project", "client", "server", "batch", "backlog"]
    group_names = {
        "ルート": "全体",
        "project": "プロジェクト",
        "client": "クライアント",
        "server": "サーバー",
        "batch": "バッチ",
        "backlog": "バックログ",
    }
    chunks = []
    for group in sorted(groups, key=lambda item: (order.index(item) if item in order else 99, item)):
        chunks.append(f"<section><h2>{html.escape(group_names.get(group, group))}</h2><ul>")
        for page in sorted(groups[group], key=lambda item: (item.source_rel != f"{group}/index.md", item.source_rel)):
            active = ' aria-current="page" class="active"' if page.source_rel == current.source_rel else ""
            href = _relative_html(current.output_rel, page.source_rel)
            search = " ".join(
                str(value)
                for value in (
                    page.doc.title,
                    page.doc.description,
                    page.doc.type,
                    " ".join(page.doc.get("tags", [])) if isinstance(page.doc.get("tags", []), list) else "",
                )
            ).lower()
            chunks.append(
                f'<li data-search="{html.escape(search, quote=True)}"><a href="{html.escape(href, quote=True)}"{active}>'
                f"{html.escape(page.doc.title)}</a></li>"
            )
        chunks.append("</ul></section>")
    return "".join(chunks)


def _render_breadcrumbs(page: Page, page_sources: set[str]) -> str:
    parts = PurePosixPath(page.source_rel).parts
    crumbs = [f'<a href="{html.escape(_relative_html(page.output_rel, "index.md"), quote=True)}">docs</a>']
    for idx, part in enumerate(parts[:-1], start=1):
        index_source = "/".join(parts[:idx]) + "/index.md"
        if index_source in page_sources:
            href = _relative_html(page.output_rel, index_source)
            crumbs.append(f'<a href="{html.escape(href, quote=True)}">{html.escape(part)}</a>')
        else:
            crumbs.append(f"<span>{html.escape(part)}</span>")
    if page.source_rel != "index.md":
        crumbs.append(f"<span>{html.escape(page.doc.title)}</span>")
    return '<nav class="breadcrumbs" aria-label="パンくず">' + '<span aria-hidden="true">/</span>'.join(crumbs) + "</nav>"


def _render_badges(doc: Any) -> str:
    type_name = doc.type or ("Index" if doc.name == "index.md" else "Log" if doc.name == "log.md" else "Document")
    status = doc.status
    trust_class, trust_label = _trust_level(doc)
    values = [
        ("type", type_name),
        (f"status-{status}", status),
        (trust_class, trust_label),
    ]
    if doc.type == "Backlog Item":
        for key in ("state", "priority", "effort", "ai"):
            value = doc.get(key)
            if value is not None:
                values.append((key, f"{key}: {value}"))
    return '<div class="badges">' + "".join(
        f'<span class="badge {html.escape(css, quote=True)}">{html.escape(str(label))}</span>'
        for css, label in values
    ) + "</div>"


def _render_related(page: Page, page_sources: set[str]) -> str:
    raw = page.doc.get("related", [])
    if not isinstance(raw, list):
        return ""
    items = []
    for value in raw:
        if not isinstance(value, str):
            continue
        rewritten, target, _warning = rewrite_href(page.source_rel, page.output_rel, value, page_sources)
        if target:
            items.append(f'<li><a href="{html.escape(rewritten, quote=True)}">{html.escape(value)}</a></li>')
    if not items:
        return ""
    return '<aside class="related"><h2>関連ドキュメント</h2><ul>' + "".join(items) + "</ul></aside>"


def _render_backlinks(page: Page, pages_by_source: dict[str, Page]) -> str:
    incoming = [source for source in pages_by_source.values() if page.source_rel in source.outbound]
    if not incoming:
        return ""
    items = []
    for source in sorted(incoming, key=lambda item: item.source_rel):
        href = _relative_html(page.output_rel, source.source_rel)
        items.append(f'<li><a href="{html.escape(href, quote=True)}">{html.escape(source.doc.title)}</a></li>')
    return '<aside class="backlinks"><h2>このページへのリンク</h2><ul>' + "".join(items) + "</ul></aside>"


def _replace_template(template: str, values: dict[str, str], source: str) -> str:
    """テンプレートだけを1回走査し、挿入した本文を再置換しない。"""
    missing: set[str] = set()

    def replace(match: re.Match[str]) -> str:
        key = match.group(1)
        if key not in values:
            missing.add(key)
            return match.group(0)
        return values[key]

    output = TEMPLATE_TOKEN_RE.sub(replace, template)
    if missing:
        raise RenderError(
            f"{source}: HTML テンプレートに未解決の項目があります: {', '.join(sorted(missing))}"
        )
    return output


def _write_atomic(path: Path, content: str) -> bool:
    data = content.replace("\r\n", "\n").encode("utf-8")
    if path.exists() and path.read_bytes() == data:
        return False
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(prefix=f".{path.name}.", suffix=".tmp", dir=str(path.parent))
    try:
        with os.fdopen(fd, "wb") as stream:
            stream.write(data)
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(tmp_name, path)
    finally:
        if os.path.exists(tmp_name):
            os.unlink(tmp_name)
    return True


def render_bundle(
    bundle: Any,
    doc_factory: Callable[[Path, Path], Any],
    output_root: Path | None = None,
    *,
    write: bool = True,
) -> RenderReport:
    """Bundle 全体を HTML 化し、結果を返す。"""
    md = _markdown()
    site_name = str(getattr(bundle, "site_name", "") or "Docs")
    actual_output_root = (output_root or bundle.root).resolve()
    source_paths = bundle.md_files(include_reserved=True)
    pages = []
    for source in source_paths:
        source_rel = source.relative_to(bundle.root).as_posix()
        pages.append(
            Page(
                doc=doc_factory(source, bundle.root),
                source_rel=source_rel,
                output_rel=str(PurePosixPath(source_rel).with_suffix(".html")),
            )
        )

    page_sources = {page.source_rel for page in pages}
    warnings: list[str] = []
    for page in pages:
        _prepare_page(page, md)
        _rewrite_page_links(page, page_sources, warnings)
        related = page.doc.get("related", [])
        if isinstance(related, list):
            for value in related:
                if isinstance(value, str):
                    _rewritten, target, warning = rewrite_href(page.source_rel, page.output_rel, value, page_sources)
                    if target:
                        page.outbound.add(target)
                    if warning:
                        warnings.append(warning)
        page.body_html = md.renderer.render(page.tokens, md.options, {})

    template = _asset_text("page.html")
    pages_by_source = {page.source_rel: page for page in pages}
    planned: list[tuple[Path, str]] = []
    for page in pages:
        target = actual_output_root.joinpath(*PurePosixPath(page.output_rel).parts)
        assets_rel = posixpath.relpath("_assets", posixpath.dirname(page.output_rel) or ".")
        long_page = len([item for item in page.headings if item.level >= 2]) >= 5 or page.char_count >= 3000
        source_hash = hashlib.sha256(page.doc.text.encode("utf-8")).hexdigest()[:16]
        source_path = bundle.root.joinpath(*PurePosixPath(page.source_rel).parts)
        source_link = os.path.relpath(source_path, target.parent).replace("\\", "/")
        type_name = page.doc.type or ("Index" if page.doc.name == "index.md" else "Log")
        values = {
            "TITLE": html.escape(page.doc.title),
            "DESCRIPTION": html.escape(
                page.doc.description or f"{page.doc.title} — {site_name} documentation"
            ),
            "SITE_NAME": html.escape(site_name),
            "LANG": "ja",
            "ASSETS": html.escape(assets_rel, quote=True),
            "SOURCE": html.escape(page.source_rel),
            "SOURCE_LINK": html.escape(source_link, quote=True),
            "HOME_LINK": html.escape(_relative_html(page.output_rel, "index.md"), quote=True),
            "SOURCE_HASH": source_hash,
            "TYPE_CLASS": html.escape(_type_class(type_name), quote=True),
            "LONG_CLASS": " long-document" if long_page else "",
            "BREADCRUMBS": _render_breadcrumbs(page, page_sources),
            "BADGES": _render_badges(page.doc),
            "NAVIGATION": _render_navigation(pages, page),
            "TOC": _render_toc(page.headings),
            "CONTENT": page.body_html,
            "RELATED": _render_related(page, page_sources),
            "BACKLINKS": _render_backlinks(page, pages_by_source),
        }
        planned.append((target, _replace_template(template, values, page.source_rel)))

    assets = {
        actual_output_root / "_assets" / "docs.css": _asset_text("docs.css"),
        actual_output_root / "_assets" / "docs.js": _asset_text("docs.js"),
    }
    written = 0
    removed = 0
    if write:
        for path, content in [*planned, *assets.items()]:
            written += int(_write_atomic(path, content))
        expected = {path.resolve() for path, _content in planned}
        for old_html in actual_output_root.rglob("*.html"):
            if old_html.resolve() in expected or not old_html.is_file():
                continue
            try:
                prefix = old_html.read_text(encoding="utf-8", errors="replace")[:256]
            except OSError:
                continue
            if OWNERSHIP_MARKER in prefix:
                old_html.unlink()
                removed += 1
    return RenderReport(len(pages), written, removed, sorted(set(warnings)), actual_output_root)

File: node/renderer.mjs
import fs from "node:fs";
import path from "node:path";
import { createHash } from "node:crypto";
import MarkdownIt from "markdown-it";
import {
  Doc,
  OkfError,
  compare,
  dataDir,
  inside,
  read,
  rel,
  write,
} from "./core.mjs";

const owned = "<!-- okf-devkit-owned:v1 -->";
const escape = (s, quotes = true) =>
  String(s)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', quotes ? "&quot;" : '"')
    .replaceAll("'", quotes ? "&#x27;" : "'");
const htmlPath = (s) => s.replace(/\.[^.\/]+$/, ".html");
const relativeHtml = (from, to) =>
  path.posix.relative(path.posix.dirname(from), htmlPath(to));
const asset = (name) => read(path.join(dataDir, "assets", name));
function* tokens(items) {
  for (const t of items) {
    yield t;
    if (t.children) yield* tokens(t.children);
  }
}
export function rewriteHref(source, output, href, sources) {
  if (!href || /^(?:#|\/\/|[a-zA-Z][a-zA-Z0-9+.-]*:)/.test(href))
    return [href, null, null];
  const match = href.match(/^([^?#]*)(\?[^#]*)?(#.*)?$/),
    raw = match?.[1];
  if (!raw || !(/\.md$/i.test(raw) || raw.endsWith("/")))
    return [href, null, null];
  let decoded;
  try {
    decoded = decodeURIComponent(raw);
  } catch {
    decoded = raw;
  }
  let target = decoded.startsWith("/")
    ? decoded.replace(/^\/+/, "")
    : path.posix.join(path.posix.dirname(source), decoded);
  if (decoded.endsWith("/")) target = path.posix.join(target, "index.md");
  target = path.posix.normalize(target);
  if (target === ".." || target.startsWith("../"))
    return [
      href,
      null,
      `${source}: Bundle 外を指すリンクを保持しました: ${href}`,
    ];
  target = target.replace(/^[./]+/, "");
  if (!sources.has(target))
    return [
      href,
      null,
      `${source}: HTML 化できないリンク先を保持しました: ${href}`,
    ];
  const relative = relativeHtml(output, target),
    encoded = encodeURI(relative).replaceAll("#", "%23").replaceAll("?", "%3F");
  return [encoded + (match[2] || "") + (match[3] || ""), target, null];
}
function markdown() {
  const md = new MarkdownIt("commonmark", {
    html: false,
    linkify: false,
    typographer: false,
  }).enable("table");
  const fence = md.renderer.rules.fence;
  md.renderer.rules.fence = (items, i, options, env, self) => {
    const t = items[i],
      language = t.info.trim().split(/\s+/)[0];
    if (language.toLowerCase() === "mermaid")
      return `<figure class="diagram"><div class="mermaid" data-mermaid-source="true">${escape(t.content, false)}</div><figcaption>Mermaid diagram</figcaption></figure>\n`;
    const rendered = fence(items, i, options, env, self);
    return (t.content.match(/\n/g) || []).length > 40
      ? `<details class="long-code"><summary>長いコードを表示（${escape(language || "code")}）</summary>${rendered}</details>\n`
      : rendered;
  };
  md.renderer.rules.table_open = () => '<div class="table-scroll"><table>\n';
  md.renderer.rules.table_close = () => "</table></div>\n";
  return md;
}
function prepare(doc, b, md) {
  const source = rel(doc.path, b.root),
    output = htmlPath(source),
    used = new Map(),
    headings = [],
    parts = [];
  const items = md.parse(
    doc.body.replace(/^\s*<!--\s*okf:auto:(?:start|end)\s*-->\s*$/gm, ""),
    {},
  );
  items.forEach((t, i) => {
    if (t.type === "heading_open" && items[i + 1]) {
      const title = items[i + 1].content.trim(),
        level = Number(t.tag.slice(1));
      const slug =
        title
          .normalize("NFKC")
          .trim()
          .toLowerCase()
          .replace(/[`*_~\[\](){}<>]/g, "")
          .replace(/[^\p{L}\p{N}_\-\u3040-\u30ff\u3400-\u9fff]+/gu, "-")
          .replace(/^[-_]+|[-_]+$/g, "") || "section";
      const count = (used.get(slug) || 0) + 1;
      used.set(slug, count);
      const anchor = slug + (count === 1 ? "" : `-${count}`);
      t.attrSet("id", anchor);
      headings.push({ title, level, anchor });
    }
    if (["inline", "fence", "code_block"].includes(t.type))
      parts.push(t.content);
  });
  return {
    doc,
    source,
    output,
    items,
    headings,
    charCount: [...parts.join("").replace(/\s+/g, "")].length,
    outbound: new Set(),
  };
}
function navigation(pages, current) {
  const groups = new Map(),
    order = ["ルート", "project", "client", "server", "batch", "backlog"];
  const names = {
    ルート: "全体",
    project: "プロジェクト",
    client: "クライアント",
    server: "サーバー",
    batch: "バッチ",
    backlog: "バックログ",
  };
  for (const page of pages) {
    const group = page.source.includes("/")
      ? page.source.split("/")[0]
      : "ルート";
    if (!groups.has(group)) groups.set(group, []);
    groups.get(group).push(page);
  }
  const rank = (g) => (order.includes(g) ? order.indexOf(g) : 99);
  return [...groups.keys()]
    .sort((a, b) => rank(a) - rank(b) || compare(a, b))
    .map((g) => {
      const items = groups
        .get(g)
        .sort(
          (a, b) =>
            Number(a.source !== `${g}/index.md`) -
              Number(b.source !== `${g}/index.md`) ||
            compare(a.source, b.source),
        );
      return (
        `<section><h2>${escape(names[g] || g)}</h2><ul>` +
        items
          .map((p) => {
            const active =
                p.source === current.source
                  ? ' aria-current="page" class="active"'
                  : "",
              tags = p.doc.get("tags", []);
            const search = [
              p.doc.title,
              p.doc.description,
              p.doc.type,
              Array.isArray(tags) ? tags.join(" ") : "",
            ]
              .join(" ")
              .toLowerCase();
            return `<li data-search="${escape(search)}"><a href="${escape(relativeHtml(current.output, p.source))}"${active}>${escape(p.doc.title)}</a></li>`;
          })
          .join("") +
        "</ul></section>"
      );
    })
    .join("");
}
function breadcrumbs(page, sources) {
  const parts = page.source.split("/"),
    crumbs = [
      `<a href="${escape(relativeHtml(page.output, "index.md"))}">docs</a>`,
    ];
  for (let i = 0; i < parts.length - 1; i++) {
    const source = parts.slice(0, i + 1).join("/") + "/index.md";
    crumbs.push(
      sources.has(source)
        ? `<a href="${escape(relativeHtml(page.output, source))}">${escape(parts[i])}</a>`
        : `<span>${escape(parts[i])}</span>`,
    );
  }
  if (page.source !== "index.md")
    crumbs.push(`<span>${escape(page.doc.title)}</span>`);
  return (
    '<nav class="breadcrumbs" aria-label="パンくず">' +
    crumbs.join('<span aria-hidden="true">/</span>') +
    "</nav>"
  );
}
function badges(doc) {
  const type =
      doc.type ||
      (doc.name === "index.md"
        ? "Index"
        : doc.name === "log.md"
          ? "Log"
          : "Document"),
    verified = doc.get("verified", []);
  const trust =
    !Array.isArray(verified) || !verified.length
      ? ["unverified", "未検証"]
      : verified.some((v) => String(v?.by || "").startsWith("human:"))
        ? ["human-reviewed", "人間レビュー済み"]
        : ["machine-confirmed", "機械確認済み"];
  const values = [["type", type], [`status-${doc.status}`, doc.status], trust];
  if (doc.type === "Backlog Item")
    for (const k of ["state", "priority", "effort", "ai"])
      if (doc.get(k) !== null) values.push([k, `${k}: ${doc.get(k)}`]);
  return (
    '<div class="badges">' +
    values
      .map(
        ([css, label]) =>
          `<span class="badge ${escape(css)}">${escape(label)}</span>`,
      )
      .join("") +
    "</div>"
  );
}
function toc(headings) {
  const visible = headings.filter((h) => h.level >= 2);
  if (visible.length < 3) return "";
  return (
    '<nav class="toc" aria-label="ページ内目次"><h2>目次</h2><ol>' +
    visible
      .map(
        (h) =>
          `<li class="toc-level-${Math.max(0, h.level - 2)}"><a href="#${escape(h.anchor)}">${escape(h.title)}</a></li>`,
      )
      .join("") +
    "</ol></nav>"
  );
}
function related(page, sources) {
  const raw = page.doc.get("related", []);
  if (!Array.isArray(raw)) return "";
  const items = raw
    .filter((v) => typeof v === "string")
    .flatMap((v) => {
      const [href, target] = rewriteHref(page.source, page.output, v, sources);
      return target
        ? [`<li><a href="${escape(href)}">${escape(v)}</a></li>`]
        : [];
    });
  return items.length
    ? '<aside class="related"><h2>関連ドキュメント</h2><ul>' +
        items.join("") +
        "</ul></aside>"
    : "";
}
function backlinks(page, pages) {
  const incoming = pages
    .filter((p) => p.outbound.has(page.source))
    .sort((a, b) => compare(a.source, b.source));
  return incoming.length
    ? '<aside class="backlinks"><h2>このページへのリンク</h2><ul>' +
        incoming
          .map(
            (p) =>
              `<li><a href="${escape(relativeHtml(page.output, p.source))}">${escape(p.doc.title)}</a></li>`,
          )
          .join("") +
        "</ul></aside>"
    : "";
}
export function renderBundle(b, output = b.root, doWrite = true) {
  output = inside(b.repo, output);
  const md = markdown(),
    pages = b.files(true).map((p) => prepare(new Doc(p, b), b, md)),
    sources = new Set(pages.map((p) => p.source)),
    warnings = [];
  for (const page of pages) {
    for (const token of tokens(page.items))
      if (token.type === "link_open") {
        const [href, target, warning] = rewriteHref(
          page.source,
          page.output,
          token.attrGet("href"),
          sources,
        );
        token.attrSet("href", href);
        if (target) page.outbound.add(target);
        if (warning) {
          token.attrJoin("class", "is-broken");
          warnings.push(warning);
        }
      }
    const refs = page.doc.get("related", []);
    if (Array.isArray(refs))
      for (const ref of refs.filter((r) => typeof r === "string")) {
        const [, target, warning] = rewriteHref(
          page.source,
          page.output,
          ref,
          sources,
        );
        if (target) page.outbound.add(target);
        if (warning) warnings.push(warning);
      }
    page.body = md.renderer.render(page.items, md.options, {});
  }
  const template = asset("page.html"),
    plan = [];
  for (const p of pages) {
    const target = inside(output, path.join(output, p.output)),
      type = p.doc.type || (p.doc.name === "index.md" ? "Index" : "Log");
    const values = {
      TITLE: escape(p.doc.title),
      DESCRIPTION: escape(
        p.doc.description || `${p.doc.title} — ${b.siteName} documentation`,
      ),
      SITE_NAME: escape(b.siteName),
      LANG: "ja",
      ASSETS: escape(
        path.posix.relative(path.posix.dirname(p.output), "_assets"),
      ),
      SOURCE: escape(p.source),
      SOURCE_LINK: escape(rel(p.doc.path, path.dirname(target))),
      HOME_LINK: escape(relativeHtml(p.output, "index.md")),
      SOURCE_HASH: createHash("sha256")
        .update(p.doc.text)
        .digest("hex")
        .slice(0, 16),
      TYPE_CLASS: escape(
        type
          .toLowerCase()
          .replace(/[^a-z0-9]+/g, "-")
          .replace(/^-|-$/g, "") || "document",
      ),
      LONG_CLASS:
        p.headings.filter((h) => h.level >= 2).length >= 5 ||
        p.charCount >= 3000
          ? " long-document"
          : "",
      BREADCRUMBS: breadcrumbs(p, sources),
      BADGES: badges(p.doc),
      NAVIGATION: navigation(pages, p),
      TOC: toc(p.headings),
      CONTENT: p.body,
      RELATED: related(p, sources),
      BACKLINKS: backlinks(p, pages),
    };
    plan.push([
      target,
      template.replace(/\{\{([A-Z0-9_]+)\}\}/g, (_m, k) => {
        if (!(k in values))
          throw new OkfError(`HTML テンプレートに未解決の項目があります: ${k}`);
        return values[k];
      }),
    ]);
  }
  const expected = new Set(plan.map(([p]) => p));
  for (const name of ["docs.css", "docs.js"])
    plan.push([
      inside(output, path.join(output, "_assets", name)),
      asset(name),
    ]);
  let written = 0,
    removed = 0;
  if (doWrite) {
    for (const [file, content] of plan) written += Number(write(file, content));
    const clean = (dir) => {
      for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
        const file = path.join(dir, e.name);
        if (e.isDirectory()) clean(file);
        else if (
          e.isFile() &&
          e.name.endsWith(".html") &&
          !expected.has(file) &&
          read(file).slice(0, 256).includes(owned)
        ) {
          fs.unlinkSync(file);
          removed++;
        }
      }
    };
    clean(output);
  }
  return {
    pages: pages.length,
    written,
    removed,
    warnings: [...new Set(warnings)].sort(compare),
    output,
  };
}
export function cmdRender(b, args) {
  const report = renderBundle(
    b,
    args.output ? path.resolve(b.repo, args.output) : b.root,
    !args.check,
  );
  if (args.hook) {
    console.log("{}");
    return 0;
  }
  report.warnings.forEach((w) => console.error(`warn: ${w}`));
  console.log(
    `HTML ${args.check ? "検証" : "生成"}: ${report.pages} ページ / 書き込み ${report.written} 件 / 削除 ${report.removed} 件 / warn ${report.warnings.length} 件 -> ${report.output}`,
  );
  return 0;
}
